test/e2e/storage/testsuites/capacity.go

```diff
@@ -0,0 +1,380 @@
+	if !ok {
+		return false, fmt.Errorf("expected *storagev1beta1.CSIStorageCapacityList, got: %T", actual)
+	}
+	h.matchingCapacities = nil
+	for _, capacity := range capacities.Items {
+		if capacity.StorageClassName == h.scName {
+			h.matchingCapacities = append(h.matchingCapacities, capacity)
+		}
+	}
+	return len(h.matchingCapacities) > 0, nil
+}
+
+func (h *haveCSIStorageCapacities) MatchedCapacities() []storagev1beta1.CSIStorageCapacity {
+	return h.matchingCapacities
+}
+
+func (h *haveCSIStorageCapacities) FailureMessage(actual interface{}) (message string) {
+	return fmt.Sprintf("no CSIStorageCapacity objects for storage class %q", h.scName)
+}
+
+func (h *haveCSIStorageCapacities) NegatedFailureMessage(actual interface{}) (message string) {
+	return fmt.Sprintf("CSIStorageCapacity objects for storage class %q:\n%s",
+		h.scName,
+		strings.Join(formatCapacities(h.matchingCapacities), "\n"),
+	)
+}
+
+// HaveCapacitiesForClassAndNodes matches objects by storage class name. It finds
+// all nodes on which the driver runs and expects one object per node.
+func HaveCapacitiesForClassAndNodes(client kubernetes.Interface, driverName, scName, topologyKey string) CapacityMatcher {
+	return &haveLocalStorageCapacities{
+		client:      client,
+		driverName:  driverName,
+		match:       HaveCapacitiesForClass(scName),
+		topologyKey: topologyKey,
+	}
+}
+
+type haveLocalStorageCapacities struct {
+	client      kubernetes.Interface
+	driverName  string
+	match       CapacityMatcher
+	topologyKey string
+
+	matchSuccess          bool
+	topologyValues        []string
+	expectedCapacities    []storagev1beta1.CSIStorageCapacity
+	unexpectedCapacities  []storagev1beta1.CSIStorageCapacity
+	missingTopologyValues []string
+}
+
+var _ CapacityMatcher = &haveLocalStorageCapacities{}
+
+func (h *haveLocalStorageCapacities) Match(actual interface{}) (success bool, err error) {
+	h.topologyValues = nil
+	h.expectedCapacities = nil
+	h.unexpectedCapacities = nil
+	h.missingTopologyValues = nil
+
+	// First check with underlying matcher.
+	success, err = h.match.Match(actual)
+	h.matchSuccess = success
+	if !success || err != nil {
+		return
+	}
+
+	// Find all nodes on which the driver runs.
+	csiNodes, err := h.client.StorageV1().CSINodes().List(context.Background(), metav1.ListOptions{})
+	if err != nil {
+		return false, err
+	}
+	for _, csiNode := range csiNodes.Items {
+		for _, driver := range csiNode.Spec.Drivers {
+			if driver.Name != h.driverName {
+				continue
+			}
+			node, err := h.client.CoreV1().Nodes().Get(context.Background(), csiNode.Name, metav1.GetOptions{})
+			if err != nil {
+				return false, err
+			}
+			value, ok := node.Labels[h.topologyKey]
+			if !ok || value == "" {
+				return false, fmt.Errorf("driver %q should run on node %q, but its topology label %q was not set",
+					h.driverName,
+					node.Name,
+					h.topologyKey)
+			}
+			h.topologyValues = append(h.topologyValues, value)
```

Comment:
If `topologyValues` is changed to a map, maybe this matcher can work for arbitrary topologies?

Reply:
Yes, I think that could work. Let me give it a try.

Reply:
I think that works. I've pushed that as a separate commit, just in case that we later find out that there was some logical flaw after all and want to revert to the original test.

Reply:
"I think" = "I tested it".

Reply:
The logical extension of that is to support multiple topology keys. But then the test code becomes as complex as the corresponding code external-provisioner and if I simply copy that code, then the same bug will be in both... let's consider this again once we actually have a storage driver with multiple keys.